ADVCONFIG.md
# Advanced Configuration

The following guide explores the variables and constants which are user-changeable within the code. The purpose of these changes is to set preferred defaults for the operation of your Proton Pack and Neutrona Wand at bootup, meaning just after the devices have battery power applied. These defaults will avoid the need to switch operation modes and other settings through the menu system.

Please consider all items as changeable at your own risk. Deviating from any recommended values may impact the operation of your pack and wand and cause unwanted or odd behavior. Bug reports or issues which arise from modification of these values will be considered moot unless accompanied by copies of changed files to aid in reproducing the behavior.

All user-configurable values reside within their respective `Configuration.h` file. Note that when loading the .ino file to the Arduino IDE this should load the related .h files automatically.

For more information on how to compile and flash a custom binary after modifying `Configuration.h`, please see [COMPILING_FLASHING.md](COMPILING_FLASHING.md) for GPStar hardware or [DIY_FLASHING.md](DIY_FLASHING.md) for DIY Arduino builds.

## EEPROM Data

For those who customise their settings and flash their boards manually, any saved settings in the Proton Pack or Neutrona Wand EEPROM can override some settings that you manually flash to your Proton Pack or Neutrona Wand's main system memory. Examples include the Cyclotron and Power Cell Delays, the number of LEDs defined for the Power Cell or Cyclotron, overheating options, etc. You can either override the EEPROM settings by setting `b_eeprom = false` in the Proton Pack and or Neutrona Wand Configuration.h file before flashing or clearing all the saved settings in your EEPROM from the EEPROM LED Menu System and EEPROM Configuration Menu System.

Refer to the [Operation Manual](OPERATION.md) for more information on how to enter the different EEPROM menu systems.

`EEPROM data saving and menu system is only available while using the GPStar Neutrona Wand board with your Neutrona Wand, or via the web interface when using an ESP32 wireless module or Attenuator.`

***The EEPROM memory has a specified life of 100,000 write/erase cycles. You will generally not reach this limit in your lifetime, but if you did or wish to disable the EEPROM loading system, refer to*** `const bool b_eeprom` ***in the Proton Pack table below.***

## Proton Pack

|Variable| Description| Default Setting |
|--------|------------|-----------------|
| uint8&#95;t i&#95;cyclotron&#95;leds | Cyclotron Lid LED count. There are 12 LEDs in the stock HasLab Cyclotron Lid. If you install a 40 LED NeoPixel ring in your Cyclotron Lid, then this should be set to 40. <br><br>`The Cyclotron Lid can support up to 40 (rgbw) addressable LEDs such as NeoPixels (ws2812b). They will change colour accordingly if you switch to video game modes firing modes.` | 12 |
| const uint16&#95;t i&#95;1984&#95;delay | This controls the delay in milliseconds when an LED changes on the Cyclotron Lid in 1984/1989 mode. The default setting is 275 which is 0.275 seconds. The 2009 video game uses 500. | 275 |
| &#35;define CYCLOTRON&#95;DELAY&#95;2021&#95;12&#95;LED | This controls the delay in milliseconds when an LED changes on the Cyclotron Lid in 2021/2024 mode when using the stock Haslab 12 LED setup. | 15 |
| &#35;define CYCLOTRON&#95;DELAY&#95;2021&#95;20&#95;LED | This controls the delay in milliseconds when an LED changes on the Cyclotron Lid in 2021/2024 mode when using a Frutto Technology 20 LED setup. | 10 |
| &#35;define CYCLOTRON&#95;DELAY&#95;2021&#95;36&#95;LED | This controls the delay in milliseconds when an LED changes on the Cyclotron Lid in 2021/2024 mode when using a Frutto Technology Max 36 LED setup. | 5 |
| &#35;define CYCLOTRON&#95;DELAY&#95;2021&#95;40&#95;LED | This controls the delay in milliseconds when an LED changes on the Cyclotron Lid in 2021/2024 mode when using a 40 LED NeoPixel ring. | 7 |
| const uint8&#95;t i&#95;1984&#95;cyclotron&#95;12&#95;leds&#95;cw&#91;4&#93; | This is the middle LED aligned in each Cyclotron lens. Put the sequence in order from lowest to highest in a clockwise direction. The top right lens is the Cyclotron lens #1 and work your way clockwise. This is for the stock Haslab LED setup. | &#123; 1, 4, 7, 10 &#125; |
| const uint8&#95;t i&#95;1984&#95;cyclotron&#95;12&#95;leds&#95;ccw&#91;4&#93; | This is the middle LED aligned in each Cyclotron lens. Put the sequence in order from lowest to highest in a counter-clockwise direction. The top right lens is the Cyclotron lens #1 and work your way counter-clockwise. This is for the stock Haslab LED setup. | &#123; 1, 10, 7, 4 &#125; |
| const uint8&#95;t i&#95;1984&#95;cyclotron&#95;20&#95;leds&#95;cw&#91;4&#93; | This is the middle LED aligned in each Cyclotron lens. Put the sequence in order from lowest to highest in a clockwise direction. The top right lens is the Cyclotron lens #1 and work your way clockwise. This is for the Frutto Technology 20 LED setup. | &#123; 2, 7, 12, 17 &#125; |
| const uint8&#95;t i&#95;1984&#95;cyclotron&#95;20&#95;leds&#95;ccw&#91;4&#93; | This is the middle LED aligned in each Cyclotron lens. Put the sequence in order from lowest to highest in a counter-clockwise direction. The top right lens is the Cyclotron lens #1 and work your way counter-clockwise. This is for the Frutto Technology 20 LED setup. | &#123; 2, 17, 12, 7 &#125; |
| const uint8&#95;t i&#95;1984&#95;cyclotron&#95;36&#95;leds&#95;cw&#91;4&#93; | This is the middle LED aligned in each Cyclotron lens. Put the sequence in order from lowest to highest in a clockwise direction. The top right lens is the Cyclotron lens #1 and work your way clockwise. This is for the Frutto Technology Max 36 LED setup. | &#123; 4, 13, 22, 31 &#125; |
| const uint8&#95;t i&#95;1984&#95;cyclotron&#95;36&#95;leds&#95;ccw&#91;4&#93; | This is the middle LED aligned in each Cyclotron lens. Put the sequence in order from lowest to highest in a counter-clockwise direction. The top right lens is the Cyclotron lens #1 and work your way counter-clockwise. This is for the Frutto Technology Max 36 LED setup. | &#123; 4, 31, 22, 13 &#125; |
| const uint8&#95;t i&#95;1984&#95;cyclotron&#95;40&#95;leds&#95;cw&#91;4&#93; | This is the middle LED aligned in each Cyclotron lens. Put the sequence in order from lowest to highest in a clockwise direction. The top right lens is the Cyclotron lens #1 and work your way clockwise. This is for the 40 NeoPixel Ring LED setup. | &#123; 0, 10, 18, 28 &#125; |
| const uint8&#95;t i&#95;1984&#95;cyclotron&#95;40&#95;leds&#95;ccw&#91;4&#93; | This is the middle LED aligned in each Cyclotron lens. Put the sequence in order from lowest to highest in a counter-clockwise direction. The top right lens is the Cyclotron lens #1 and work your way counter-clockwise. This is for the 40 NeoPixel Ring LED setup. | &#123; 0, 28, 18, 10 &#125; |
| bool b&#95;clockwise | Set to true for the Cyclotron lights to spin clockwise. Set to false to be counter clockwise. This can be controlled by an optional switch on pin 29.<br><br>`This can be toggled from the Neutrona Wand sub menu system.` | true |
| bool b&#95;fade&#95;cyclotron&#95;led | When set to false, the LEDs on the Cyclotron Lid in 1984/1989 mode will not fade between lens transitions which is closer to their appearance in the films. Set to true to get an appearance closer to the 2009 video game. <br><br>`This can be toggled from the Neutrona Wand sub menu system.` | true |
| bool b&#95;cyclotron&#95;single&#95;led | When set to true, the LEDs on the Cyclotron Lid will light up the middle LED only per Cyclotron lens in 1984 and 1989 mode. When set to false, 3 LEDs per lens will be displayed instead. <br><br>`This can be toggled from the Neutrona Wand sub menu system.` | true |
| bool b&#95;cyclotron&#95;simulate&#95;ring | When set to true, using LEDs that are not a ring will simulate a ring rotation for the Cyclotron LEDs in the lid. This only affects Afterlife and Frozen Empire modes (2021/2024).<br><br>`This has no effect if you are using a 40 pixel ring in your Cyclotron Lid.`| true |
| bool b&#95;cyclotron&#95;colour&#95;toggle | If you are using Cyclotron Lid LEDs and Inner Cyclotron LEDs with RGB support, such as the Frutto Technology Cyclotron LEDs or NeoPixel Rings etc. You can toggle if you want it to change colours to match the Video Game Modes or stay the default Red at all times. The default setting is true, which makes the Cyclotron Lid and Inner Cyclotron change colours to match the Video Game Modes.<br><br>`This has no affect on the stock HasLab Cyclotron Lid LEDs, which are red only. This can be toggled in the Neutrona Wand sub menu system.` | true |
| const bool b&#95;cyclotron&#95;haslab&#95;chsv&#95;colour&#95;change | When using 12 LEDs for the Cyclotron Lid (Stock HasLab), the system will default it to always red because changing the colour space with a CHSV colour object affects the brightness slightly for non RGB pixels such as the ones used in the HasLab Cyclotron Lid. If you are using 12 LEDs that are RGB, they will always be red when this is set to false. Setting this to true will override it and make the LEDs change colour if they are RGB.<br><br>`This only affects when you are using 12 LEDs for the Cyclotron Lid (HasLab)` | false |
| uint8&#95;t i&#95;powercell&#95;leds | Power Cell LED count. There are 13 LEDs in the stock HasLab Power Cell. If you install a Frutto Technology Power Cell which has 15 LEDs, then change this to 15. You can also add your own addressable LEDs, adjust the count as necessary.<br><br>`If you add more Power Cell LEDs, you may want to adjust the i_powercell_delay_2021 and i_powercell_delay_1984 to lower numbers to speed up the Power Cell delay to reflect the additional LEDs. For 15 LEDs, 1984 at 60 and 2021 at 34 are good settings.` | 13 |
| &#35;define POWERCELL&#95;DELAY&#95;1984&#95;13&#95;LED | This controls the delay in milliseconds when an LED changes on the Power Cell in 1984/1989 mode when using a 13 LED Power Cell. | 46 |
| &#35;define POWERCELL&#95;DELAY&#95;2021&#95;13&#95;LED | This controls the delay in milliseconds when an LED changes on the Power Cell in 2021/2024 mode when using a 13 LED Power Cell. | 40 |
| &#35;define POWERCELL&#95;DELAY&#95;1984&#95;15&#95;LED | This controls the delay in milliseconds when an LED changes on the Power Cell in 1984/1989 mode when using a 15 LED Power Cell. | 40 |
| &#35;define POWERCELL&#95;DELAY&#95;2021&#95;15&#95;LED | This controls the delay in milliseconds when an LED changes on the Power Cell in 2021/1989 mode when using a 15 LED Power Cell. | 34 |
| uint8&#95;t i&#95;powercell&#95;delay&#95;1984 | This is the actual control variable for the Power Cell LED delay in 1984/1989 mode. It is recommended to change the above defines and only change this variable to the appropriate define for your LED configuration. | POWERCELL&#95;DELAY&#95;1984&#95;13&#95;LED |
| uint8&#95;t i&#95;powercell&#95;delay&#95;2021 | This is the actual control variable for the Power Cell LED delay in 2021/2024 mode. It is recommended to change the above defines and only change this variable to the appropriate define for your LED configuration. | POWERCELL&#95;DELAY&#95;2021&#95;13&#95;LED |
| bool b&#95;powercell&#95;invert | Set to true to invert the Power Cell animation.| false |
| bool b&#95;powercell&#95;colour&#95;toggle | If you are using Power Cell LEDs with RGB support, such as the Frutto Technology Power Cells, you can toggle if you want it to change colours to match the Video Game Modes or stay the default Blue at all times. The default setting is true, which makes the Power Cell change colours to match the Video Game Modes.<br><br>`This has no affect on the stock HasLab Power Cell LEDs, which are blue only. This can be toggled in the Neutrona Wand sub menu system.` | true |
| uint8&#95;t i&#95;inner&#95;cyclotron&#95;cake&#95;num&#95;leds | This setting controls how many optional LEDs are in the Inner (cake) Cyclotron. Default setting of 35 is for a 35 pixel NeoPixel ring. If you change this for a smaller NeoPixel ring, you may need to alter the i&#95;2021&#95;inner&#95;delay and i&#95;1984&#95;inner&#95;delay to lower values to slow down the rotation speed for a smaller ring. | 35 |
| LED&#95;TYPES CAVITY&#95;LED&#95;TYPE | Sets the default LED type for the inner cyclotron cavity fairy lights. Choices are `RGB_LED`, `GRB_LED` or `GBR_LED`. | `RGB_LED` |
| uint8&#95;t i&#95;inner&#95;cyclotron&#95;cavity&#95;num&#95;leds | You can use up to 20 LEDs. This is generally for adding additional LEDs inside the cyclotron cavity to create spark effects or anything else. | 0 |
| LED&#95;TYPES CAKE&#95;LED&#95;TYPE | Sets the default LED type for the inner cyclotron cake. Choices are `RGB_LED`, `GRB_LED` or `GBR_LED`. | `GBR_LED` |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;1984&#95;12&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 1984/1989 mode when using a 12 LED NeoPixel ring. | 15 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;2021&#95;12&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 2021/2024 mode when using a 12 LED NeoPixel ring. | 12 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;1984&#95;23&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 1984/1989 mode when using a 23 LED NeoPixel ring. | 12 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;2021&#95;23&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 2021/2024 mode when using a 23 LED NeoPixel ring. | 8 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;1984&#95;24&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 1984/1989 mode when using a 24 LED NeoPixel ring. | 12 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;2021&#95;24&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 2021/2024 mode when using a 24 LED NeoPixel ring. | 8 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;1984&#95;26&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 1984/1989 mode when using a 26 LED NeoPixel ring. | 12 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;2021&#95;26&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 2021/2024 mode when using a 26 LED NeoPixel ring. | 8 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;1984&#95;35&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 1984/1989 mode when using a 35 LED NeoPixel ring. | 9 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;2021&#95;35&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 2021/2024 mode when using a 35 LED NeoPixel ring. | 5 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;1984&#95;36&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 1984/1989 mode when using a 36 LED NeoPixel ring. | 9 |
| &#35;define INNER&#95;CYCLOTRON&#95;DELAY&#95;2021&#95;36&#95;LED | This controls the delay in milliseconds when an LED changes on the Inner Cyclotron (cake) in 2021/2024 mode when using a 36 LED NeoPixel ring. | 5 |
| uint8&#95;t i&#95;1984&#95;inner&#95;delay | This is the actual control variable for the Inner Cyclotron (cake) LED delay in 1984/1989 mode. It is recommended to change the above defines and only change this variable to the appropriate define for your LED configuration. | INNER&#95;CYCLOTRON&#95;DELAY&#95;1984&#95;35&#95;LED |
| uint8&#95;t i&#95;2021&#95;inner&#95;delay | This is the actual control variable for the Inner Cyclotron (cake) LED delay in 2021/2024 mode. It is recommended to change the above defines and only change this variable to the appropriate define for your LED configuration. | INNER&#95;CYCLOTRON&#95;DELAY&#95;2021&#95;35&#95;LED |
| uint8&#95;t i&#95;spectral&#95;powercell&#95;custom&#95;colour |  This is the default colour of the Power Cell when in the Spectral Custom firing mode. The default colour is purple, and this ranges from 1 to 254.<br><br>`This firing mode is part of the video game firing modes.` `This colour can be adjusted from the EEPROM LED menu.` | 200 |
| uint8&#95;t i&#95;spectral&#95;cyclotron&#95;custom&#95;colour |  This is the default colour of the Cyclotron when in the Spectral Custom firing mode. The default colour is purple, and this ranges from 1 to 254.<br><br>`This firing mode is part of the video game firing modes.` `This colour can be adjusted from the EEPROM LED menu.` | 200 |
| uint8&#95;t i&#95;spectral&#95;cyclotron&#95;inner&#95;custom#95;colour |  This is the default colour of the Inner Cyclotron (cake) when in the Spectral Custom firing mode. The default colour is purple, and this ranges from 1 to 254.<br><br>`This firing mode is part of the video game firing modes.` `This colour can be adjusted from the EEPROM LED menu.` | 200 |
| uint8&#95;t i&#95;spectral&#95;powercell&#95;custom&#95;saturation |  This is the saturation of the Power Cell when in the Spectral Custom firing mode. This ranges from 1 to 254.<br><br>`This firing mode is part of the video game firing modes.` `This saturation can be adjusted from the EEPROM LED menu.` | 254 |
| uint8&#95;t i&#95;spectral&#95;cyclotron&#95;custom&#95;saturation |  This is the saturation of the Cyclotron when in the Spectral Custom firing mode. This ranges from 1 to 254.<br><br>`This firing mode is part of the video game firing modes.` `This saturation can be adjusted from the EEPROM LED menu.` | 254 |
| uint8&#95;t i&#95;spectral&#95;cyclotron&#95;inner&#95;custom&#95;saturation |  This is the saturation of the Inner Cyclotron (cake) when in the Spectral Custom firing mode. This ranges from 1 to 254.<br><br>`This firing mode is part of the video game firing modes.` `This saturation can be adjusted from the EEPROM LED menu.` | 254 |
| uint8&#95;t i&#95;powercell&#95;brightness | The default brightness of your Power Cell LEDs in percentages. 0 = off, 100 = maximum brightness.<br><br>`This can be toggled from the Neutrona Wand sub menu system.` | 100 |
| uint8&#95;t i&#95;cyclotron&#95;brightness | The default brightness of your Cyclotron LEDs in percentages. 0 = off, 100 = maximum brightness.<br><br>`This can be toggled from the Neutrona Wand sub menu system.` | 100 |
| uint8&#95;t i&#95;cyclotron&#95;inner&#95;brightness | The default brightness of your Inner Cyclotron LEDs in percentages. 0 = off, 100 = maximum brightness.<br><br>`This can be toggled from the Neutrona Wand sub menu system.` | 100 |
| uint8&#95;t i&#95;cyclotron&#95;panel&#95;brightness | The default brightness of your Frutto Inner Cyclotron LED Switch Panel (if installed) in percentages. 0 = off, 100 = maximum brightness.<br><br>`This can be toggled from the Neutrona Wand sub menu system.` | 100 |
| const uint8&#95;t STARTUP&#95;VOLUME | The default startup master volume of the Proton Pack. This is a percentage value of between 100% and 0%. 100 = highest, 0 = lowest. When a Neutrona Wand is connected, it will sync to this value. | 100 |
| const uint8&#95;t STARTUP&#95;VOLUME&#95;MUSIC | The default music volume of the Proton Pack. This is a percentage value of between 100% and 0%. 100 = highest, 0 = lowest. When a Neutrona Wand is connected, it will sync to this value. | 100 |
| const uint8&#95;t STARTUP&#95;VOLUME&#95;EFFECTS | The default sound effects volume of the Proton Pack. This is a percentage value of between 100% and 0%. 100 = highest, 0 = lowest. When a Neutrona Wand is connected, it will sync to this value. | 100 |
| const int8&#95;t MINIMUM&#95;VOLUME | You can set the minimum lowest volume the Proton Pack can go. 0 = loudest, and -70 = quietest. | -60 |
| const uint8&#95;t VOLUME&#95;MULTIPLIER | Percentage increments of master volume change. | 5 |
| const uint8&#95;t VOLUME&#95;MUSIC&#95;MULTIPLIER | Percentage increments of the music volume change. | 5 |
| const uint8&#95;t VOLUME&#95;EFFECTS&#95;MULTIPLIER | Percentage increments of the sound effects volume change. | 5 |
| const bool b&#95;onboard&#95;amp&#95;enabled | Set to true to enable the onboard amplifier of the WAV Trigger. Turning off the onboard amp draws less power. If using the AUX cable jack of the WAV Trigger, the amp can be disabled to save power. If you are using the output pins directly on the WAV Trigger board to your speakers, you will need to enable the amp. <br><br>`The onboard mono audio amplifier and speaker connector specifications: 2w into 4 Ohms, 1.25W into 8 Ohms`| false |
| bool b&#95;demo&#95;light&#95;mode | When set to true, the Proton Pack will turn on automatically when it receives power. If you want your Proton Pack to be silent, change your STARTUP&#95;VOLUME to be 0 or unplug the power to your amplifier. | false |
| bool b&#95;stream&#95;effects | When set to true, various impact and other stream effects will overlap and mix randomly into the Proton Stream for an added experience.<br><br>`This can be toggled from the Neutrona Wand sub menu system` | true |
| bool b&#95;overheat&#95;strobe | If you want the optional N-Filter NeoPixel jewel to strobe during overheat, set to true. When false, the light stays solid white during overheat. | false |
| bool b&#95;overheat&#95;lights&#95;off | When the pack is overheating, the Power Cell will ramp down and turn off. The Cyclotron also turns off. When set to false, the Power Cell and Cyclotron stay on during overheat but ramps down and spins very slowly. | true |
| bool b&#95;overheat&#95;sync&#95;to&#95;fan | When set to true, smoke from the N-Filter will only operate at the same time the fan is operating. When set to false (default), smoke in the N-Filter will pump earlier than the fan to fill up the N-Filter with some smoke. <br><br>If you have a smoke/fan kit for the N-Filter that operates the smoke and fan at the same time, and you are connected to the smoke N-Filter pin on the pack board, then you would want to set this to true.<br><br>If you have a smoke kit where the smoke and fan are independently connected to the pack board, setting to false is preferred. | false |
| bool b&#95;smoke&#95;enabled | Enable or disable smoke settings. This can be toggled with a switch on PIN 37. <br><br>`This can be toggled from the Neutrona Wand sub menu system.` | true |
| bool b&#95;smoke&#95;1&#95;continuous&#95;firing | Set to true to enable smoke 1 during continuous firing events. <br><br>`This can be overridden if b_smoke_enabled is set to false` | true |
| bool b&#95;smoke&#95;2&#95;continuous&#95;firing | Set to true to enable smoke 2 during continuous firing events. <br><br>`This can be overridden if b_smoke_enabled is set to false` | true |
| bool b&#95;fan&#95;continuous&#95;firing | Set to true to enable the fan during continuous firing events. <br><br>`This can be overridden if b_smoke_enabled is set to false` | true |
| bool b&#95;fan&#95;booster&#95;continuous&#95;firing | Set to true to enable the fan in the booster tube during continuous firing events. <br><br>`This can be overridden if b_smoke_enabled is set to false` | true |
| const bool b&#95;smoke&#95;continuous&#95;level&#95;1 | Set to true to enable smoke events in `wand power level 1` during continuous firing events.<br><br>`This can be overridden if b_smoke_enabled is set to false or if all 3 of the smoke_1, smoke_2 and fan_continuous are set to false` | true |

HISTORY.md
# Project History

Work began in December of 2022 by Michael Rajotte through regular posts on the GBFans forum. Initial efforts focused on creation of a "cheat device" or more commonly referred to as a "keepalive mod" which would defeat the automatic power-down feature built into the Habro Haslab equipment. By January of 2023 most of the worldwide community were receiving their orders from Hasbro and interest in such cheat devices was growing. At this stage Michael was successful in creating a working mod though it required soldering to delicate components within the Neutrona Wand which was outside the skillset of many end-users.

Though once a minimal keepalive solution was found, Michael began experimenting with other features such as cyclotron lights and audio upgrades. Conversations with Jonathan Gunn about potential features exposed some obvious limitations with the Haslab equipment, and so the idea of a microcontroller replacement was born. By late January the wand had a working prototype using an Arduino Nano to control the built-in hardware and work began on the pack controller. Within weeks the pack was running its core functions from an Arduino as well and both devices could communicate with each other for effects and operations. Interest from the community was growing, as was the list of requests for features.

Progress updates continued on the GBFans forums as new prototyping work began in the background through February. The introduction of the GitHub page for all source code and project files was introduced. Work also began to prototype and beta test Michael's schematics in the community. By early March the experimental hardware and software was replicated by Dustin Grau and efforts began to formally document instructions for the community. The first true replacement to the stock Haslab controllers was on its way to adoption but only by the most-skilled users.

Michael continued adding features and modding his equipment to accommodate community requests. Documentation turned to schematics and video demonstrations, but was still a daunting prospect to most modders. Then in April a PCB schematic was created by Jonathan which changed everything--the idea would be to combine the electronic components which lived on protoboards onto a single board powered by a ATmega2560 chip. Several revisions were rapidly created and modified, and within weeks a prototype board for the pack was created and available. By May, both the pack and wand had viable prototypes which offered the same socket connections as their replaced originals, along with user-friendly options for attaching the wand components which were originally soldered to the stock controller.

This was the breakthrough moment for the project. Handwritten notes turned into documentation, which turned into manufactured circuits, then gave way to a means of creating drop-in controllers that would be accessible to modders with a wider range of skills. Even those with no soldering abilities could pick up a screwdriver and install these components. At this stage Angie Rutto of Frutto Technology began working with Michael to identify add-on parts that could integrate with the controllers in use and a bevvy of new options became available. Replacement lights for the powercell and cyclotron lid allowed for full-color effects on the pack, while a new bargraph unit allowed for a plug-and-play upgrade for the wand. The project was now becoming a platform with supported accessories.

Efforts then shifted from prototyping and revisional changes to potential worldwide distribution. Polls and preorders gauged the public interest and efforts began to turn components into kits. Michael, Jonathan, and Angie took on additional roles as regional distributors covering the EU, North America, and the UK (respectively). Jonathan created the first website for ordering and modders were able to submit their choice for base kits, special upgrade kits, and optional add-ons.


source/Attenuator/Settings.h
/**
 *   GPStar Attenuator - Ghostbusters Proton Pack & Neutrona Wand.
 *   Copyright (C) 2023 Michael Rajotte <michael.rajotte@gpstartechnologies.com>
 *                    & Dustin Grau <dustin.grau@gmail.com>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, see <https://www.gnu.org/licenses/>.
 *
 */

#pragma once

const char SETTINGS_page[] PROGMEM = R"=====(
<!DOCTYPE html>
<html>
<head>
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
  <title>Equipment Settings</title>
  <link rel="icon" href="data:;base64,iVBORw0KGgo=">
  <link rel="stylesheet" href="/style.css">
</head>
<body>
  <h1>Modify Equipment Settings</h1>
  <div class="block">
    <p>
      Change system configuration options using the available toggles/selectors.
      <br/>
      Options can only be changed when the pack and wand are not powered/running!
    </p>
    <br/>

    <h1>System Options</h1>
    <div class="block left">
      <b>System Mode:</b>
      <select id="systemMode" name="systemMode">
        <option value="0">Super Hero</option>
        <option value="1">Mode Original</option>
      </select>
      </br>
      <b>Theme Year:</b>
      <select id="themeYear" name="themeYear">
        <option value="1">System Toggle</option>
        <option value="2">1984</option>
        <option value="3">1989</option>
        <option value="4">Afterlife</option>
        <!-- <option value="5">Frozen Empire</option> -->
      </select>
      </br>
      <b>Default Master Volume:</b>
      <input type="range" id="masterVol" name="masterVol" min="0" max="100" value="100" step="2"/>
      </br>
      <b>Proton Stream Impact Effects:</b>
      <label class="switch">
        <input id="streamEffects" name="streamEffects" type="checkbox">
        <span class="slider round"></span>
      </label>
      </br>
      <b>Smoke Effects:</b>
      <label class="switch">
        <input id="smokeEffects" name="smokeEffects" type="checkbox">
        <span class="slider round"></span>
      </label>
      </br>
      <b>Strobe N-Filter on Overheat:</b>
      <label class="switch">
        <input id="overheatStrobe" name="overheatStrobe" type="checkbox">
        <span class="slider round"></span>
      </label>
      </br>
      <b>Lights Off During Overheat:</b>
      <label class="switch">
        <input id="overheatLightOff" name="overheatLightOff" type="checkbox">
        <span class="slider round"></span>
      </label>
      </br>
      <b>Overheat Sync to Fan:</b>
      <label class="switch">
        <input id="overheatSync" name="overheatSync" type="checkbox">
        <span class="slider round"></span>
      </label>
      </br>
      <b>Demo Light Mode:</b>
      <label class="switch">
        <input id="demoEnabled" name="demoEnabled" type="checkbox">
        <span class="slider round"></span>
      </label>
    </div>

    <h1>Cyclotron Lid</h1>
    <div class="block left">
      <b>LED Count:</b>
      <select id="cycCount" name="cycCount">
        <option value="12">12 - Stock</option>
        <option value="20">20 - Frutto</option>
        <option value="40">40 - Ring</option>
      </select>
      <br/>
      <b>Spectral Custom Color (Hue):</b>
      <input type="range" id="cycHue" name="cycHue" min="2" max="254" value="254" step="2"/>
      <br/>
      <b>Spectral Custom Saturation:</b>
      <input type="range" id="cycSat" name="cycSat" min="2" max="254" value="254" step="2"/>
      <br/>
      <b>Direction:</b>
      <select id="cycDir" name="cycDir">
        <option value="0">Counter-Clockwise</option>
        <option value="1">Clockwise</option>
      </select>
      </br>
      <b>Center LEDs:</b>
      <select id="cycCenter" name="cycCenter">
        <option value="0">3 LED</option>
        <option value="1">1 LED</option>
      </select>
      </br>
      <b>Enable Video Game Colors:</b>
      <label class="switch">
        <input id="vgCyclotron" name="vgCyclotron" type="checkbox">
        <span class="slider round"></span>
      </label>
      <br/>
      <b>Simulate Ring:</b>
      <label class="switch">
        <input id="cycRing" name="cycRing" type="checkbox">
        <span class="slider round"></span>
      </label>
    </div>

    <h1>Cyclotron Cake</h1>
    <div class="block left">
      <b>LED Count:</b>
      <select id="cakeCount" name="cakeCount">
        <option value="12">12</option>
        <option value="23">23</option>
        <option value="24">24</option>
        <option value="35">35</option>
      </select>
      <br/>
      <b>Spectral Custom Color (Hue):</b>
      <input type="range" id="cakeHue" name="cakeHue" min="2" max="254" value="254" step="2"/>
      <br/>
      <b>Spectral Custom Saturation:</b>
      <input type="range" id="cakeSat" name="cakeSat" min="2" max="254" value="254" step="2"/>
      <br/>
      <b>Swap Red/Green LEDs (GRB):</b>
      <label class="switch">
        <input id="cakeGRB" name="cakeGRB" type="checkbox">
        <span class="slider round"></span>
      </label>
    </div>

    <h1>Power Cell</h1>
    <div class="block left">
      <b>LED Count:</b>
      <select id="pcCount" name="pcCount">
        <option value="13">13 - Stock</option>
        <option value="15">15 - Frutto</option>
      </select>
      <br/>
      <b>Spectral Custom Color (Hue):</b>
      <input type="range" id="pcHue" name="pcHue" min="2" max="254" value="254" step="2"/>
      <br/>
      <b>Spectral Custom Saturation:</b>
      <input type="range" id="pcSat" name="pcSat" min="2" max="254" value="254" step="2"/>
      </br>
      <b>Enable Video Game Colors:</b>
      <label class="switch">
        <input id="vgPowercell" name="vgPowercell" type="checkbox">
        <span class="slider round"></span>
      </label>
    </div>

    <br/>
    <a href="/">&laquo; Back</a>
    &nbsp;&nbsp;&nbsp;
    <button type="button" class="green" onclick="saveSettings()">Save</button>
  </div>

  <script type="application/javascript">
    window.addEventListener("load", onLoad);

    function onLoad(event) {
      getStatus();
      getPreferences();
    }

    function getPreferences() {
      var xhttp = new XMLHttpRequest();
      xhttp.onreadystatechange = function() {
        if (this.readyState == 4 && this.status == 200) {
          console.log(JSON.parse(this.responseText));
        }
      };
      xhttp.open("GET", "/preferences", true);
      xhttp.send();
    }

    function getStatus() {
      var xhttp = new XMLHttpRequest();
      xhttp.onreadystatechange = function() {
        if (this.readyState == 4 && this.status == 200) {
          // Use the current settings to determine what's configured.
          var data = JSON.parse(this.responseText);
        }
      };
      xhttp.open("GET", "/status", true);
      xhttp.send();
    }

    function saveSettings() {
      var body = JSON.stringify({name: value});

      var xhttp = new XMLHttpRequest();
      xhttp.onreadystatechange = function() {
        if (this.readyState == 4 && this.status == 200) {
          var jObj = JSON.parse(this.responseText);
          alert(jObj.status); // Always display status returned.
        }
      };
      xhttp.open("PUT", "/settings/update", true);
      xhttp.setRequestHeader("Content-Type", "application/json");
      xhttp.send(body);
    }
  </script>
</body>
</html>
)=====";